### ISSUE TITLE
[eCBIS] Remove the family member

### ISSUE THREAD

**Issue by dubdabasoduba** [opensrp/fhircore#767]
**Name of feature to enhance**
- We would like to have the ability to delete a family member.
- A family is currently represented by the patient resources of the head of household.

**Description of feature**
- [ ] Add a `Remove Person` button on the family member profile.
- [ ] On clicking the button, launch a confirmation dialog to confirm the delete actions.
- [ ] If yes is clicked, then update the family member `active` attribute and set it to `false`. 

**Remove the head of family**
- [ ] Figure out how to list the family head as the first entry of the family member list
- [ ] If delete is initiated we should have a popup/new page to let the CHW select the new head of household. 
- [ ] After the selection is made, show a confirmation dialog. 
- [ ] If yes, then update the family member `active` attribute and set it to false. 
- [ ] Remove the `family` tag from the patient resource. 
- [ ] Add the `family tag` to the patient resource of the newly selected family head. 

![edit info](https://user-images.githubusercontent.com/4817531/142628729-13c15d36-638e-46a9-97ee-85e8653bad2b.png)

**Comment by pld** [comment]
If they're not already, can we refactor all the family related logic to a module in the engine?

**Comment by dubdabasoduba** [comment]
> they're not already, can we refactor all the family related logic to a module in the engine?

https://github.com/opensrp/fhircore/discussions/764

**Comment by maimoonak** [comment]
Remove Family Member:
- Accidental entry and want to DELETE. (Hard [delete](http://hl7.org/fhir/http.html#delete) or soft delete (mark as inactive)?)
- If removal is due to the reason that it is no more applicable for given scenario i.e. No more family member, moving this person to other family would sort out the issue
- If it is due to the fact that person has died we need to set death date and mark as inactive
- Shall we show inactive i.e. soft deleted people into members`s list as striked off record or may be disabled or in a different color.

**Comment by maimoonak** [comment]
Remove the Family Head

Again why someone wants to delete Family Head
- Accidental entry
- Death of Head
- Change of Head

- By current design Family Head info is listed into card and not into list of members. 
- If delete is initiated we should have a popup/new page to let the CHW select the new head of household. Shall we only allow new head from existing members, OR someone else from other family can become head of a current family?
- If yes, then update the family member active attribute and set it to false (It depends if person is just stepping out of Head position and be member only now. In this case this person would refer to new family head)

**Comment by pld** [comment]
After delete will we be able to track that someone was in the past assigned as head of household? that'll be useful (and maybe in some situations required) for auditing

**Comment by maimoonak** [comment]
We create Flag for pregnancy and family. We need to mark flag as inactive and set period end date. https://hl7.org/fhir/flag.html

**Comment by f-odhiambo** [comment]
@maimoonak Does this implementation cascade to the ANC list too? i.e If removed from the Family member list are you also removed from the ANC register?  Do we add a prompt to make sure you confirm what you are doing?

@dubdabasoduba Is there a scenario that should allow you to return a member if this is initiated by accident?

**Comment by maimoonak** [comment]
_Does this implementation cascade to the ANC list too_? (No. Currently we are not doing it, but we can if needed)
_Do we add a prompt to make sure you confirm what you are doing_? (Yes. A prompt would be there.)
_Do we blur the user in the list? perhaps ? or disappear completely in list_? (What shall we do? If its permanent delete we can not do anything, if it is death we can show it, if family member is moved out its reference would be changed and we can not show it in list.. so in short only incase of death we can show a member despite removal )

**Comment by pld** [comment]
I think normally we hide deceased and there's an option somewhere to unhide them, CC @rowo

**Comment by f-odhiambo** [comment]
@maimoonak Will get back to you on this, I will schedule a call with some more folks e.g. @dubdabasoduba @ndegwamartin @ellykits to get a preliminary solution for this

**Comment by f-odhiambo** [comment]
**Some notes here** 

**Death of a Family member ** 
1. Mark patient as deceased and inactive 
2. Remove them from the register 
3. Mark any active CarePlan as status  - Revoked 

**Entered in Error of a Family member**
1. Remove from the Family 
2. if ANC client retain in ANC register and remove from the family list 
3. If not ANC then remove from the Family list and just remain a Patient in HAPI 

**Moved out - family member**  - Moving from resident and starting a new family 
1. This is an edge case 
2. There needs to be a check to be able  to 1st create a new family and then remove it from the previous family list  
3. The person becomes a new family HHH
4. Can be further discussed and scoped 

**Death of a HHH** 
1. Mark patient as deceased and inactive 
2. Remove them from the register 
3. Mark any active CarePlan as status  - Revoked 
4. Assign new HHH
5. Link the family members to new HHH

**Entered in Error of a HHH**
1. Assign new HHH
2. Link the family members to new HHH
3. Link the previous HH CarePlan to the new HHH


**Moved out - HHH**  - Moving from resident and starting a new family 
1. Change address 


**Assigning new HHH**
1. Open Question  - Should we assign a new member or replace it from the Existing family member list?
2. In the interim select from existing family members 


CC @pld @dubdabasoduba @ellykits Kindly review and give your feedback

**Comment by dubdabasoduba** [comment]
> **Some notes here**
> 
> **Death of a Family member**
> 
> 1. Mark patient as deceased and inactive
> 2. Remove them from the register
> 3. Change CarePlan status from _Active_ to _Revoked_
> 
> **Entered in Error of a Family member**
> 
> 1. Remove from the Family list
> 2. if ANC client, retain in ANC register and remove from the family list
> 3. If not ANC client,  then remove from the Family list and just retain them as  a Patient in HAPI (How doe we retrieve them in future ?)
> 
> **Moved out - family member** - Moving from resident and starting a new family
> 
> 1. This is an edge case
> 2. There needs to be a check to be able  to 1st create a new family and then remove him/her from the previous family list
> 3. The person becomes a new family HHH
>    _To be further discussed and scoped_
> 
> **Death of a HHH**
> 
> 1. Mark patient as deceased and inactive
> 2. Remove them from the register
> 3. Mark any active CarePlan as status  - Revoked
> 4. Assign new HHH: Open Question  - Should we assign a new member or replace it from the Existing family member list?
> 5. Link the family members to new HHH
> 6. Link the previous Household CarePlan to the new HHH
> 
> **Entered in Error of a HHH**
> 
> 1. Assign new HHH
> 2. Link the family members to new HHH
> 3. Link the previous Household CarePlan to the new HHH
> 
> **Moved out - HHH** - Moving from resident and starting a new family
> 
> 1. Change address
> 
> **Assigning new HHH**
> 
> 1. Open Question  - Should we assign a new member or replace it from the Existing family member list?
> 2. In the interim select from existing family members
> 
> CC @pld @dubdabasoduba @ellykits Kindly review and give your feedback

**Comment by dubdabasoduba** [comment]
**Some notes here**

**Death of a Family member**

> 1. Mark patient as deceased and inactive
> 2. Remove them from the register

Currently, we use tags to show which registers a user is in. This will mean removing all the register tags from the user. 

> 3. Change CarePlan status from _Active_ to _Revoked_

I don't think we use CarePlans currently.  Unless this is for the future. 

> 
> **Entered in Error of a Family member**
> 
> 1. Remove from the Family list
> 2. if ANC client, retain in ANC register and remove from the family list
> 3. If not ANC client,  then remove from the Family list and just retain them as  a Patient in HAPI (How do we retrieve them in the future ?)

We might need to use the OpenSRP FHIR web our a way to show this patient by some form of search. 

> 
> **Moved out - family member** - Moving from resident and starting a new family
> 
> 1. This is an edge case
> 2. There needs to be a check to be able  to 1st create a new family and then remove him/her from the previous family list

I think we could skip this step and move to 3 directly. Step 2 could just be a dialog explaining what will happen. 
> 3. The person becomes a new family HHH
>    _To be further discussed and scoped_
> 
> **Death of a HHH**
> 
> 1. Mark patient as deceased and inactive
> 2. Remove them from the register

Currently, we use tags to show which registers a user is in. This will mean removing all the register tags from the user. 
> 3. Mark any active CarePlan as status  - Revoked

This is for later when we have careplans are implemented
> 4. Assign new HHH: Open Question  - Should we assign a new member or replace it from the Existing family member list?

Do we register a new family member if the existing ones aren't selected? 
> 5. Link the family members to new HHH
> 6. Link the previous Household CarePlan to the new HHH

This is for later when we have careplans are implemented
> 
> **Entered in Error of a HHH**
> 
> 1. Assign new HHH
> 2. Link the family members to new HHH
> 3. Link the previous Household CarePlan to the new HHH

This is for later when we have careplans are implemented
> 
> **Moved out - HHH** - Moving from resident and starting a new family
> 
> 1. Change address
> 
> **Assigning new HHH**
> 
> 1. Open Question  - Should we assign a new member or replace it from the Existing family member list?

Show a list of all family members to select from. Also, have a button to create a new family member to add a new one
> 2. In the interim select from existing family members
> 
> CC @pld @dubdabasoduba @ellykits Kindly review and give your feedback

NB: - The whole of this documentation works on the implementation of having patients linked to the HHH through the `link` attribute on the patient resource.

**Comment by dubdabasoduba** [comment]
> We create Flag for pregnancy and family. We need to mark flag as inactive and set period end date. https://hl7.org/fhir/flag.html

Does this mean we sync the patients then, when trying to load them to a register we check the flags to know the registers they belong to? 

```
[
  {
    "system": "https://www.snomed.org",
    "code": "35359004",
    "display": "Family"
  },
  {
    "system": "http://loinc.org",
    "code": "LA15173-0",
    "display": "Pregnant"
  }
]
```
If that's the case what are these use for?

For the tags do, we add the patients to the `subject` attribute?

**Comment by dubdabasoduba** [comment]
> After delete will we be able to track that someone was in the past assigned as head of household? that'll be useful (and maybe in some situations required) for auditing

@pld What vision do you have here? Do we just track that the person was a HHH or do we need the extra information like which family (list of patients) were in the family, when he was removed as the HHH

**Comment by dubdabasoduba** [comment]
> @maimoonak Does this implementation cascade to the ANC list too? i.e If removed from the Family member list are you also removed from the ANC register? Do we add a prompt to make sure you confirm what you are doing? Do we blur the user in the list? perhaps ? or disappear completely in list?
> 
> @dubdabasoduba Is there a scenario that should allow you to return a member if this is initiated by accident?

@f-odhiambo I think from a UX perspective, deletes have prompts to make sure the user clearly knows what they are about to do and the consequences of doing so. I would recommend we have it too. 

I think we still need to give a provision to re-add the patient to the family. 

My thinking from a UX perspective
- have the Add Familly member button open up internal/advanced search first. this will allow searching of all patients using `name` or `identifier`. If we find them then we can add them back. 
- if we don't find them, then we open the registration form.

**Comment by AnnieMungai** [comment]
@dubdabasoduba @f-odhiambo @pld @rowo 
In current implementations, here are some things to consider:
- When a family member is deceased, they are removed from all registers but are visible on the family list but marked as deceased and user cannot tap on them. So their record is viewable but is inactive (mostly greyed out).
- A date of death should also be considered for any person removed by death. This will allow computations in the app for the age of death (important reporting indicator). 

**Things to consider:** 
- When removing a family member, currently this person is completely removed from a HH. (Assumption is that they will be registered in th app under the place they moved to). This ensures that we do not double-report anyone. So when implementing the removing a person from a HH, consider how we report this without counting them more than once i.e, ensuring they are only counted in 1 location. 

Question:
- If a family is tied to a HHH, what happens if a HHH is dead and the next person available does not qualify to be HHH? Should the user be allowed to remove such a HHH? Example - a child under 16years does not qualify to be a HHH. How does this affect how this HH is handled when the HHH is removed?

**Comment by pld** [comment]
> > After delete will we be able to track that someone was in the past assigned as head of household? that'll be useful (and maybe in some situations required) for auditing
> 
> @pld What vision do you have here? Do we just track that the person was a HHH or do we need the extra information like which family (list of patients) were in the family, when he was removed as the HHH

backing up, I'm thinking of this from a data it'd be interesting to be able to calculate perspective, like:
- how many changes in HHH does the average family have over X time period?
- does health quality suffer/change right before or after a change in HHH?
- does health visit attendance change right before or after a change in HHH?

tbc, definitely doesn't need to be easy to calculate these types of things but we'd like to have the raw data to make it possible.

with that context, logging an event for assigned, unassigned HHH, family member added, removed to HH would let a data scientist recreate stats for the above from structured log data

also, tbc, this is very much in the nice-to-have category, but bringing it up so that if LOE is not a factor but more about how we architect, we're taking it into account

**Comment by pld** [comment]
> 
> Question:
> 
>     * If a family is tied to a HHH, what happens if a HHH is dead and the next person available does not qualify to be HHH? Should the user be allowed to remove such a HHH? Example - a child under 16years does not qualify to be a HHH. How does this affect how this HH is handled when the HHH is removed?

Good question, guess the related question is can you have an HH w/o an HHH? Sounds like the answer should be yes, since you've given a clear situation where that occurs.

**Comment by maimoonak** [comment]
> > We create a Flag for pregnancy and family. We need to mark the flag as inactive and set the period end date. https://hl7.org/fhir/flag.html
> 
> @maimoonak Does this mean we sync the patients then, when trying to load them to a register we check the flags to know the registers they belong to?
> 
> ```
> [
>   {
>     "system": "https://www.snomed.org",
>     "code": "35359004",
>     "display": "Family"
>   },
>   {
>     "system": "http://loinc.org",
>     "code": "LA15173-0",
>     "display": "Pregnant"
>   }
> ]
> ```
> 
> If that's the case what are these use for? They are part of the tags for this [patient](https://turn-fhir.smartregister.org/fhir/Patient/127f5741-9710-4206-8ce7-606752e8eb2d)
> 
> For the tags do, we add the patients to the `subject` attribute? We need to discuss further the use of flags since it might conflict with the usage of [groups/careteams in syncing using location mapping](https://github.com/opensrp/fhircore/discussions/746).

@dubdabasoduba No. Nothing in code is dependent on Flag so far. When adding Family or Pregnancy we do create Tag into Patient. The same Tag.code is used in Flag as well. The Flag was created to keep the historic data as Flag has period(start and end) which allows to track when a Patient was entitled for that specific condition and when it has been ended. Tag represents current state and can be removed when Patient no longer qualifies, but Flag can show multiple events which led to highlight Patient for that.

**Comment by maimoonak** [comment]
We also use [Condition](https://hl7.org/fhir/condition.html) instead of [Flag](https://hl7.org/fhir/flag.html). Both of them are timebound.

**Comment by dubdabasoduba** [comment]
@AnnieMungai 

> Question:
> - If a family is tied to a HHH, what happens if a HHH is dead and the next person available does not qualify to be HHH? Should the user be allowed to remove such a HHH? Example - a child under 16years does not qualify to be a HHH. How does this affect how this HH is handled when the HHH is removed?

- We would mark the head of household as deceased but the family would still remain and the deceased family member would still remain as the HHH until the living family members are of the right age to be HHHs 

Questions
- Do we have a scenario where we would want to create a family without a HHH e.g registering a household after the HHH passed on and during registration, none of the living family members are eligible to be an HHH. 

cc @pld

**Comment by HenryRae** [comment]
@dubdabasoduba the actions in this [list](https://github.com/opensrp/fhircore/issues/767#issue-1057769582) screenshot should be sorted according to priority/actionable items.
i.e am thinking it would be:
1. edit info
2. pregnancy outcome
3. add vitals
4. add conditions
5. BMI widget
6. view past encounters
7. view family
8. remove this person - _which should actually state "Remove member"_

**Comment by HenryRae** [comment]
My thoughts/adding on [this](https://github.com/opensrp/fhircore/issues/767#issuecomment-983392696) are:

**Death of a Family member**
1. Mark patient as deceased and inactive
_if a patient is being removed, we request the user to input a reason... so you can provide some options we know like moving out of community, or starting new family or deceased etc, and thus we won't have a long list of dropdown on individual profile screen with various options and if user is dead, then we can do as stated in 2nd bullet_ 
2. Show them on the family list but mark them as deceased and disabled ✅
3. Change CarePlan status from Active to Revoked ✅

For members **entered in error**, we included toasters(with undo action but disappears after x-time) when a member is added to a family. this can be a solution to this. And if a user missed out on this action, then we can include it as an option in the remove member reasons.
And am not sure this use case will exist if a member can be entered in error then added as an ANC patient before CHA realises the member was entered in error.

And if the **removed member(by reason provided)** is the family head, then we request the user to assign new family head before continuing or make this optional(thus CHA can skip assigning a new family head) - but this raises another **question**, does the family name stay as it was or it changes to new surname of the new family head?

**Entered in error of a HHH** - should we allow users to change family head at any point of time? are we currently doing this?

Let me know what you think,
Good thought @f-odhiambo .
cc: @rowo

**Comment by dubdabasoduba** [comment]
> @dubdabasoduba the actions in this [list](https://github.com/opensrp/fhircore/issues/767#issue-1057769582) screenshot should be sorted according to priority/actionable items. i.e am thinking it would be:
> 
> 1. edit info
> 2. pregnancy outcome
> 3. add vitals
> 4. add conditions
> 5. BMI widget
> 6. view past encounters
> 7. view family
> 8. remove this person - _which should actually state "Remove member"_
> 
> cc: @f-odhiambo @rowo

@HenryRae We would like a dialog/confirmation popup for some of the actions like `Remove this person` Is this something we can get a design for or do we go ahead and use the default android provided ones

cc: @ellykits

**Comment by dubdabasoduba** [comment]
> My thoughts/adding on [this](https://github.com/opensrp/fhircore/issues/767#issuecomment-983392696) are:
> 
> **Death of a Family member**
> 
> 1. Mark patient as deceased and inactive
>    _if a patient is being removed, we request the user to input a reason... so you can provide some options we know like moving out of community, or starting new family or deceased etc, and thus we won't have a long list of dropdown on individual profile screen with various options and if user is dead, then we can do as stated in 2nd bullet_
> 2. Show them on the family list but mark them as deceased and disabled white_check_mark
> 3. Change CarePlan status from Active to Revoked white_check_mark
> 
> For members **entered in error**, we included toasters(with undo action but disappears after x-time) when a member is added to a family. this can be a solution to this. And if a user missed out on this action, then we can include it as an option in the remove member reasons. And am not sure this use case will exist if a member can be entered in error then added as an ANC patient before CHA realises the member was entered in error.
> 
> And if the **removed member(by reason provided)** is the family head, then we request the user to assign new family head before continuing or make this optional(thus CHA can skip assigning a new family head) - but this raises another **question**, does the family name stay as it was or it changes to new surname of the new family head?
> 
> **Entered in error of a HHH** - should we allow users to change family head at any point of time? are we currently doing this?
> 
> Let me know what you think, Good thought @f-odhiambo . cc: @rowo

I think I agree with providing reasons for various actions.  @f-odhiambo I think this is where we can plug in the encounter pieces you mentioned.

**Comment by dubdabasoduba** [comment]
@HenryRae Also when showing the deceased family members does it make sense from a UX/UI perspective to have them at the bottom of the member list? 

I think we should also not include them in the list of available family members when selecting a new HHH.

**Comment by HenryRae** [comment]
Yes, @dubdabasoduba it would be good to move deceased family members to the bottom of the list and it's not ideal to include them to the list of available family members(well since they are no longer available, RIP) when selecting a new HHH.

I will create designs and think about the flow of [this](https://github.com/opensrp/fhircore/issues/767#issuecomment-985001679) dialog.
cc: @rowo

**Comment by dubdabasoduba** [comment]
> **Some notes here** This is based on discussions with @dubdabasoduba @ellykits and @ndegwamartin Nov 30th 2021
> 
> **Death of a Family member**
> 
> 1. Mark patient as deceased and inactive
> 2. Show them on the family list but mark them as deceased and disabled
> 3. Change CarePlan status from _Active_ to _Revoked_
> 
> **Entered in Error of a Family member**
> 
> 1. Remove from the Family list
> 2. if ANC client, retain in ANC register and remove from the family list
> 3. If not ANC client,  then remove from the Family list and just retain them as  a Patient in HAPI (How do we retrieve them in future ?) Advanced search?
> 
> **Moved out - family member** - Moving from resident and starting a new family
> 
> 1. There needs to be a check to be able  to 1st create a new family and then remove him/her from the previous family list
> 2. The person becomes a new family HHH
>    _To be further discussed and scoped_
> 
> **Death of a HHH**
> 
> 1. Mark patient as deceased and inactive
> 2. Show them on the family list but mark them as deceased and disabled
> 3. Mark any active CarePlan as status  - Revoked
> 4. Assign new HHH: Open Question  - Should we assign a new member or replace it from the Existing family member list?
> 5. Link the family members to new HHH
> 6. Link the previous Household CarePlan to the new HHH
> 
> **Entered in Error of a HHH**
> 
> 1. Assign new HHH
> 2. Link the family members to new HHH
> 3. Link the previous Household CarePlan to the new HHH
> 
> **Moved out - HHH** - Moving from resident and starting a new family
> 
> 1. Change address
> 
> **Assigning new HHH**
> 
> 1. Open Question  - Should we assign a new member or replace it from the Existing family member list?
> 2. In the interim select from existing family members
> 
> CC @pld @dubdabasoduba @ellykits Kindly review and give your feedback

@maimoonak Does this unblock you to start the implementation?
Do you need us to look at things for better scoping?

**Comment by maimoonak** [comment]
@dubdabasoduba Nopes. The above discussion with Francis was clear and current implementation follows it, except one case where it was changing family address incase of moveout (basically incase of HHH its just a change of address and not come under removal)

**Comment by maimoonak** [comment]
**Remove Family Member** (does not allow removal of Head or already inactive member)
- Reason popup
  - Entered in Error 
    - active=false, 
    - clear head reference, 
    - revoke active Careplan, 
    - revoke active Condition and Flag (we need to make sure that in future if devs using Episode of care, Appointment etc as source of truth, we ensure the status updates to those as well or consider patient status)
  - Moved out -> 
    - clear head reference
    
**Death Family Member OR Head** (does not allow already deleted or dead member)
- Death Date picker
  - Revoke active Careplan (make sure to omit family careplans)
  - Revoke active Condition, Flag

**Change Family Head**
- Pick new head (alive members excluding head)
   - Assign family tag to new head
   - Remove family tag from older head
   - Remove old head link from new head
   - Assign new head reference to old head
   - Assign new head reference to all members
   - Assign address to new head
   - Assign family care plans to new head
   - Add Family Flag for new head
   - Mark old Flag as inactive


<img width="200" height="400"  src="https://user-images.githubusercontent.com/4829880/145032800-a086d23e-1b70-4b0f-8577-139604f3afe8.jpg"  /> <img width="200" height="400"  src="https://user-images.githubusercontent.com/4829880/145032808-5a3201ac-c380-4b7f-a2d3-da8f67c94be2.jpg"  /> <img width="200" height="400"  src="https://user-images.githubusercontent.com/4829880/145032821-6edbeb77-7874-42e3-8aeb-ecba93fcb06c.jpg"  /> 
<img width="200" height="400"  src="https://user-images.githubusercontent.com/4829880/145032840-402eb975-99dd-478c-ae6f-db391b9212b5.jpg"  /> <img width="200" height="400"  src="https://user-images.githubusercontent.com/4829880/145032855-07d1ba4a-71fc-4403-a27e-c50de64d3e3c.jpg"  />

**Comment by f-odhiambo** [comment]
**Update Dec 14th** 
1. Failing CI test cases
2. Failing Engines tests  for  Measure Evaluate  - I think due to the OutOfMemoryIssue reason the MeasureEvaluate test failing

**Comment by pld** [comment]
closed by https://github.com/opensrp/fhircore/pull/838
